plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveExecutorUtil.java

```diff
@@ -0,0 +1,140 @@
+     * <p>
+     * Note: using this method allows simple parallelization of tasks within executor, when sub-tasks
+     * are also scheduled in that executor, without risking starvation when pool is saturated.
+     *
+     * @throws ExecutionException if any task fails; exception cause is the first task failure
+     */
+    public static <T> List<T> processWithAdditionalThreads(Collection<Callable<T>> tasks, Executor executor)
+            throws ExecutionException
+    {
+        List<Task<T>> wrapped = tasks.stream()
+                .map(Task::new)
+                .collect(toImmutableList());
+        CompletionService<TaskResult<T>> completionService = new ExecutorCompletionService<>(executor);
+        List<Future<?>> futures = new ArrayList<>(wrapped.size());
+        Context tracingContext = Context.current();
+
+        // schedule in the executor
+        for (int i = 0; i < wrapped.size(); i++) {
+            int index = i;
+            Task<T> task = wrapped.get(i);
+            futures.add(completionService.submit(() -> {
+                if (!task.take()) {
+                    return null; // will be ignored
+                }
+                try (var _ = tracingContext.makeCurrent()) {
+                    return new TaskResult<>(index, task.callable.call());
+                }
+            }));
+        }
+
+        try {
+            // process in the calling thread (in reverse order, as an optimization)
+            List<T> results = new ArrayList<>(nCopies(wrapped.size(), null));
+            int pending = wrapped.size();
+            for (int i = wrapped.size() - 1; i >= 0; i--) {
+                // process ready results to fail fast on exceptions
+                for (Future<TaskResult<T>> ready = completionService.poll(); ready != null; ready = completionService.poll()) {
+                    TaskResult<T> taskResult = ready.get();
+                    if (taskResult != null) {
+                        results.set(taskResult.taskIndex(), taskResult.result());
+                        pending--;
+                    }
+                }
+                Task<T> task = wrapped.get(i);
+                if (!task.take()) {
+                    continue;
+                }
+                try {
+                    results.set(i, task.callable.call());
+                    pending--;
+                }
+                catch (Exception e) {
+                    throw new ExecutionException(e);
+                }
+            }
+
+            while (pending > 0) {
+                TaskResult<T> taskResult = completionService.take().get();
+                if (taskResult != null) {
+                    results.set(taskResult.taskIndex(), taskResult.result());
+                    pending--;
+                }
+            }
+
+            return results;
+        }
+        catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            futures.forEach(future -> future.cancel(true));
```

Comment:
Move this to a finally block to ensure the futures are always cancled, so we don't do any unnecessary work (for example when we throw ExecutionException above)

Reply:
good catch, thanks!